File: Classes/WMISet/WMISet/Examples/delphi/ProcessControl/FrmSelectComputerU.pas
unit FrmSelectComputerU;

interface
{$I define.inc}

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, ExtCtrls, WmiProcessControl;

type
  TFrmSelectComputer = class(TForm)
    cmbComputers: TComboBox;
    lblSelectComputer: TLabel;
    pnlCredentials: TPanel;
    chbConnectAsCurrent: TCheckBox;
    lblDomain: TLabel;
    edtDomain: TEdit;
    lblUserName: TLabel;
    edtUserName: TEdit;
    Label1: TLabel;
    edtPassword: TEdit;
    btnOk: TButton;
    btnCancel: TButton;
    procedure chbConnectAsCurrentClick(Sender: TObject);
    procedure cmbComputersDropDown(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    { Private declarations }
    FListLoaded: boolean;
  public
    { Public declarations }
  end;

var
  FrmSelectComputer: TFrmSelectComputer;

implementation

{$R *.dfm}

procedure TFrmSelectComputer.chbConnectAsCurrentClick(Sender: TObject);
  procedure DisableEdit(AEdit: TEdit);
  begin
    AEdit.Enabled   := false;
    AEdit.Color     := clBtnFace;
    AEdit.Text      := '';
  end;

  procedure EnableEdit(AEdit: TEdit);
  begin
    AEdit.Enabled   := true;
    AEdit.Color     := clWindow;
  end;
  
begin
  if chbConnectAsCurrent.Checked then
  begin
    DisableEdit(edtDomain);
    DisableEdit(edtUserName);
    DisableEdit(edtPassword);
  end else
  begin
    EnableEdit(edtDomain);
    EnableEdit(edtUserName);
    EnableEdit(edtPassword);
  end;
end;

procedure TFrmSelectComputer.cmbComputersDropDown(Sender: TObject);
var
  vControl: TWmiProcessControl;
  vCursor: TCursor;
begin
  if not FListLoaded then
  begin
    vControl := TWmiProcessControl.Create(nil);
    vCursor  := Screen.Cursor;
    Screen.Cursor := crHourGlass;
    try
      vControl.ListServers(cmbComputers.Items);
      if cmbComputers.Items.IndexOf(vControl.SystemInfo.ComputerName) = - 1 then
        cmbComputers.Items.Add(vControl.SystemInfo.ComputerName);
      FListLoaded := true;
    finally
      Screen.Cursor := vCursor;
      vControl.Free;
    end;
  end;
end;

procedure TFrmSelectComputer.FormCreate(Sender: TObject);
begin
  FListLoaded := false;
end;

end.

File: Classes/WMISet/WMISet/Examples/cbuilder/ProcessControl/FrmSelectComputerU.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "FrmSelectComputerU.h"
#include "WmiAbstract.hpp"
#include "WmiComponent.hpp"
#include "WmiProcessControl.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TFrmSelectComputer *FrmSelectComputer;
//---------------------------------------------------------------------------
__fastcall TFrmSelectComputer::TFrmSelectComputer(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TFrmSelectComputer::chbConnectAsCurrentClick(TObject *Sender)
{
  if (chbConnectAsCurrent->Checked) {
    DisableEdit(edtDomain);
    DisableEdit(edtUserName);
    DisableEdit(edtPassword);
  } else {
    EnableEdit(edtDomain);
    EnableEdit(edtUserName);
    EnableEdit(edtPassword);
  };
}

//---------------------------------------------------------------------------
void __fastcall TFrmSelectComputer::DisableEdit(TEdit *AEdit) {
    AEdit->Enabled   = false;
    AEdit->Color     = clBtnFace;
    AEdit->Text      = "";
};

void __fastcall TFrmSelectComputer::EnableEdit(TEdit *AEdit) {
    AEdit->Enabled   = true;
    AEdit->Color     = clWindow;
};
void __fastcall TFrmSelectComputer::cmbComputersDropDown(TObject *Sender)
{
  if (!FListLoaded) {
    TWmiProcessControl *vControl = new TWmiProcessControl(NULL);
    TCursor vCursor = Screen->Cursor;
    Screen->Cursor = crHourGlass;
    __try {
      vControl->ListServers(cmbComputers->Items);
      if (cmbComputers->Items->IndexOf(vControl->SystemInfo->ComputerName) == -1) {
        cmbComputers->Items->Add(vControl->SystemInfo->ComputerName);
      }  
      FListLoaded = true;
    } __finally {
      Screen->Cursor = vCursor;
      vControl->Free();
    };
  };
}
//---------------------------------------------------------------------------
void __fastcall TFrmSelectComputer::FormCreate(TObject *Sender)
{
  FListLoaded = false;
}
//---------------------------------------------------------------------------
